File: src/viberoom/engine/cache.py
"""Caching: disk cache for rendered previews, in-memory LRU for decoded
linear arrays (decode is the expensive step when iterating on a recipe)."""

from __future__ import annotations

import hashlib
import io
import os
from collections import OrderedDict
from pathlib import Path

import numpy as np
from PIL import Image

from viberoom.engine.decode import decode_linear
from viberoom.engine.pipeline import render
from viberoom.recipe.schema import Recipe


#: How much decoded imagery to keep resident, in MB. A half-size decode of a
#: 24 MP RAW is ~70 MB, so the default holds a filmstrip's worth; full-size
#: decodes are ~290 MB and a couple will fit. Override for small machines.
DECODE_CACHE_MB = int(os.environ.get("VIBEROOM_DECODE_CACHE_MB", "1024"))


class DecodeCache:
    """LRU of decoded linear ndarrays, bounded by total bytes.

    Bounding by *entry count* was the original design and it made browsing
    pathological: two entries meant a third image evicted the first, so paging
    back and forth through a filmstrip re-ran LibRaw every single time. Sizes
    here vary by two orders of magnitude between a half-size JPEG and a
    full-size RAW, so bytes are the only budget that means anything.
    """

    def __init__(self, max_bytes: int | None = None):
        self._max_bytes = max_bytes if max_bytes is not None else DECODE_CACHE_MB * 1024 * 1024
        self._store: OrderedDict[tuple, np.ndarray] = OrderedDict()
        self._bytes = 0

    def _key(self, path: Path, half_size: bool) -> tuple:
        # st_mtime_ns rather than st_mtime: the float loses sub-microsecond
        # resolution, so a file rewritten twice inside one tick can collide.
        return (str(path), path.stat().st_mtime_ns, half_size)

    def get(self, path: Path, half_size: bool) -> np.ndarray:
        key = self._key(path, half_size)
        if key in self._store:
            self._store.move_to_end(key)
            return self._store[key]

        arr = decode_linear(path, half_size=half_size)
        self.put(key, arr)
        return arr

    def put(self, key: tuple, arr: np.ndarray) -> None:
        # Entries are handed out by reference, so an in-place op anywhere in
        # the pipeline would silently poison every later render of the same
        # file. Making them read-only turns that from a corrupted image into
        # an immediate ValueError at the offending line.
        arr.flags.writeable = False
        if key in self._store:
            self._bytes -= self._store.pop(key).nbytes
        self._store[key] = arr
        self._bytes += arr.nbytes
        # Always keep the entry just inserted, even if it alone busts the
        # budget — evicting it would mean decoding it again immediately.
        while self._bytes > self._max_bytes and len(self._store) > 1:
            self._bytes -= self._store.popitem(last=False)[1].nbytes

    def clear(self) -> None:
        self._store.clear()
        self._bytes = 0

    @property
    def resident_mb(self) -> float:
        return self._bytes / (1024 * 1024)


_decode_cache = DecodeCache()


# bump when pipeline math changes so cached previews re-render
# 4: previews render at preview resolution rather than decode resolution
PIPELINE_VERSION = 4


def preview_cache_key(path: Path, recipe: Recipe, size: int) -> str:
    raw = f"v{PIPELINE_VERSION}|{path}|{path.stat().st_mtime}|{recipe.canonical_json()}|{size}"
    return hashlib.sha1(raw.encode()).hexdigest()


#: Extra input resolution kept beyond what the output strictly needs, so that
#: rotation and perspective have real pixels to resample from at the edges
#: instead of inventing them.
_RESAMPLE_MARGIN = 1.15


def _preview_scale(linear: np.ndarray, recipe: Recipe, size: int) -> float:
    """How much to shrink the decoded frame before rendering it.

    A crop is the reason this is not simply `size / long_edge`: cropping to a
    quarter of the frame means the surviving quarter has to carry the whole
    output, so four times the input is needed to hold detail.

    Returns 1.0 when the frame is already small enough — never upscales.
    """
    crop = recipe.geometry.crop
    kept = min(max(crop.right - crop.left, 1e-3), 1.0), min(
        max(crop.bottom - crop.top, 1e-3), 1.0
    )
    needed = size / min(kept) * _RESAMPLE_MARGIN
    return min(1.0, needed / max(linear.shape[0], linear.shape[1]))


def _downscale_linear(linear: np.ndarray, scale: float) -> np.ndarray:
    """Shrink a linear float32 image, resampling per channel with Lanczos.

    Done in linear light, which is where averaging pixels is physically
    meaningful — the same reduction applied after the sRGB transfer curve
    would darken every gradient.
    """
    h, w = linear.shape[:2]
    th, tw = max(1, round(h * scale)), max(1, round(w * scale))
    if (th, tw) == (h, w):
        return linear

    # PIL's float mode is single-channel, so the channels go one at a time.
    out = np.empty((th, tw, linear.shape[2]), dtype=np.float32)
    for c in range(linear.shape[2]):
        plane = Image.fromarray(np.ascontiguousarray(linear[..., c]), mode="F")
        # reducing_gap lets PIL box-reduce most of the way before the Lanczos
        # pass, which is where nearly all of this function's time went.
        out[..., c] = np.asarray(
            plane.resize((tw, th), Image.LANCZOS, reducing_gap=3.0), dtype=np.float32
        )
    # Lanczos overshoots at hard edges; negative light is not a thing.
    return np.clip(out, 0, None, out=out)


def render_preview(path: Path, recipe: Recipe, size: int, cache_dir: Path) -> bytes:
    """Rendered-with-recipe JPEG preview, disk-cached."""
    key = preview_cache_key(path, recipe, size)
    cached = cache_dir / f"{key}.jpg"
    if cached.exists():
        return cached.read_bytes()

    linear = _decode_cache.get(path, half_size=True)

    # Render at roughly the size actually being asked for. This used to run
    # the whole pipeline at decode resolution and shrink the result at the
    # very end, which meant a 400 px grid thumbnail cost exactly as much as a
    # 1600 px loupe view — the expensive part being every op in between.
    scale = _preview_scale(linear, recipe, size)
    rgb = render(_downscale_linear(linear, scale), recipe, scale)

    im = Image.fromarray(rgb)
    im.thumbnail((size, size), Image.LANCZOS)
    buf = io.BytesIO()
    im.save(buf, "JPEG", quality=90)
    data = buf.getvalue()
    cached.write_bytes(data)
    return data


def render_full(path: Path, recipe: Recipe) -> np.ndarray:
    """Full-resolution render for export (no disk cache).

    Goes through the decode cache: exporting several variants or formats of
    one image used to re-run LibRaw for each, which is by far the most
    expensive step. Safe now that the cache is bounded by bytes rather than
    entries — a ~290 MB full decode can be admitted and evicted sensibly.
    """
    linear = _decode_cache.get(path, half_size=False)
    return render(linear, recipe)


def render_full_float(path: Path, recipe: Recipe) -> np.ndarray:
    """Full-resolution float [0,1] render for high-bit-depth export."""
    from viberoom.engine.pipeline import render_float

    linear = _decode_cache.get(path, half_size=False)
    return render_float(linear, recipe)

File: frontend/src/local/decode/cache.ts
/** LRU of decoded linear frames, bounded by total bytes.
 *
 * Port of `engine/cache.py::DecodeCache`, and bounded the same way for the same
 * reason: entry-count bounding made browsing pathological, because two entries
 * meant a third image evicted the first and paging back and forth through a
 * filmstrip re-ran LibRaw every single time. Frame sizes vary by two orders of
 * magnitude between a small JPEG and a full-size RAW, so bytes are the only
 * budget that means anything.
 *
 * It matters more in a tab than it did on the server. A 24 MP decode is ~290 MB
 * of float32 and a browser will kill the tab rather than swap, so the budget is
 * a fraction of the server's default.
 */

import type { LinearImage } from './types.ts'

/** Full-resolution decodes, which are enormous: 197 MB for a Canon CR3, 524 MB
 *  for a Nikon Z7, 698 MB for a Sony ARW. Only one is ever wanted at a time —
 *  a full decode exists to produce a preview-sized frame and is not needed
 *  again until export — so this is sized to hold roughly one, and the
 *  keep-what-was-just-inserted rule does the rest. */
const DEFAULT_MAX_BYTES = 384 * 1024 * 1024

/** Preview-sized frames, which are small: a 2048px frame is ~32 MB. These are
 *  what makes going back to a photo instant, so they get the larger share and
 *  a separate budget.
 *
 *  Sharing one budget with the full decodes was the bug: producing a single
 *  preview allocated several hundred megabytes and evicted every frame worth
 *  keeping, so arrowing back through a filmstrip re-ran LibRaw every time. */
const FRAME_MAX_BYTES = 512 * 1024 * 1024

export class DecodeCache {
  private store = new Map<string, LinearImage>()
  private bytes = 0
  private maxBytes: number

  constructor(maxBytes: number = DEFAULT_MAX_BYTES) {
    this.maxBytes = maxBytes
  }

  /** Decoded frame for `key`, running `decode` only on a miss.
   *
   * In-flight decodes are not deduplicated here — the caller owns request
   * lifetimes (an abandoned loupe navigation should not keep a decode alive),
   * and a duplicate decode is merely slow, whereas a shared promise that
   * outlives its abort is a leak. */
  async get(key: string, decode: () => Promise<LinearImage>): Promise<LinearImage> {
    const hit = this.store.get(key)
    if (hit) {
      // Map preserves insertion order, so re-inserting is the whole of "touch".
      this.store.delete(key)
      this.store.set(key, hit)
      return hit
    }
    const img = await decode()
    this.put(key, img)
    return img
  }

  put(key: string, img: LinearImage): void {
    const existing = this.store.get(key)
    if (existing) {
      this.store.delete(key)
      this.bytes -= existing.data.byteLength
    }
    this.store.set(key, img)
    this.bytes += img.data.byteLength
    // Always keep the entry just inserted, even if it alone busts the budget —
    // evicting it would mean decoding it again immediately.
    while (this.bytes > this.maxBytes && this.store.size > 1) {
      const oldest = this.store.keys().next().value as string
      this.bytes -= this.store.get(oldest)!.data.byteLength
      this.store.delete(oldest)
    }
  }

  clear(): void {
    this.store.clear()
    this.bytes = 0
  }

  get residentMb(): number {
    return this.bytes / (1024 * 1024)
  }
}

export const decodeCache = new DecodeCache()

/** Preview-sized frames. Separate from `decodeCache` on purpose — see the
 *  budgets above. */
export const frameCache = new DecodeCache(FRAME_MAX_BYTES)

/**
 * Cache identity for a file.
 *
 * `lastModified` is in here for the same reason the server keys on mtime: an
 * edited file reusing a stale decode is a wrong picture, not a stale one. Size
 * disambiguates the same-name-same-tick case that a millisecond timestamp
 * cannot.
 */
export function fileKey(file: File, suffix = ''): string {
  return `${file.name}|${file.size}|${file.lastModified}${suffix}`
}
